### PR TITLE
Adding a gRPC python client for vtgate / Vitess.

### PR DESCRIPTION
**Pull request by alainjobart**
It passes python_client_test so far, not more.
Stuck on the batch API for now, not sure we need the dual-API.

@aaijazi if you're curious. Not ready to pull in yet.

### CODE REVIEW

**Comment by alainjobart** [comment]
@aaijazi @enisoc @dumbunny 
This is now ready for review. I still have lots of clean-up to do, but they'll be in another batch.

**Comment by enisoc** [comment]
LGTM

File: py/vtdb/grpc_vtgate_client.py [hunk 1]
```diff
@@ -0,0 +1,536 @@
+# Copyright 2013 Google Inc. All Rights Reserved.
+# Use of this source code is governed by a BSD-style license that can
+# be found in the LICENSE file.
+
+"""A simple, direct connection to the vttablet query server, using gRPC.
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: "vtgate proxy server" instead of "vttablet query server"?

File: py/vtdb/grpc_vtgate_client.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+  def dial(self):
+    if self.stub:
+      self.stub.close()
+
+    p = urlparse('http://' + self.addr)
+    channel = implementations.insecure_channel(p.hostname, p.port)
+    self.stub = vtgateservice_pb2.beta_create_Vitess_stub(channel)
+
+  def close(self):
+    self.stub = None
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: good idea to rollback the existing session here, if there is one?

File: py/vtdb/grpc_vtgate_client.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+  def cursor(self, *pargs, **kwargs):
+    cursorclass = kwargs.pop('cursorclass', None) or vtgate_cursor.VTGateCursor
+    return cursorclass(self, *pargs, **kwargs)
+
+  def begin(self, effective_caller_id=None):
+    try:
+      request = vtgate_pb2.BeginRequest()
+      _add_caller_id(request, effective_caller_id)
+      response = self.stub.Begin(request, self.timeout)
+      self.effective_caller_id = effective_caller_id
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: A short comment on why you're storing effective_call_id in an instance variable would be helpful here

File: py/vtdb/grpc_vtgate_client.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+  kwargs_as_str = vtgate_utils.convert_exception_kwargs(kwargs)
+  exc.args += args
+  if kwargs_as_str:
+    exc.args += kwargs_as_str,
+  new_args = (type(exc).__name__,) + exc.args
+  if isinstance(exc, vtgate_utils.VitessError):
+    new_exc = exc.convert_to_dbexception(new_args)
+  elif isinstance(exc, face.ExpirationError):
+    new_exc = dbexceptions.TimeoutError(new_args)
+  elif isinstance(exc, face.AbortionError):
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: What exactly is an AbortionError? When are they returned by gRPC?

File: py/vtdb/grpc_vtgate_client.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+      _add_caller_id(request, self.effective_caller_id)
+      self._add_session(request)
+      self.stub.Rollback(request, self.timeout)
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      raise _convert_exception(e)
+    finally:
+      self.session = None
+      self.effective_caller_id = None
+
+  def _execute(self, sql, bind_variables, keyspace_name, tablet_type,
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: This should have a more detailed docstring, maybe just copy the one from the GoRPC client?

File: py/vtdb/grpc_vtgate_client.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+
+def _add_key_ranges(request, keyranges):
+  for kr in keyranges:
+    encoded_kr = request.key_ranges.add()
+    encoded_kr.start = kr.Start
+    encoded_kr.end = kr.End
+
+
+def _extract_rpc_error(exec_method, error):
+  if error.code:
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: Wait, are you sure this is doing the right thing?

_extract_rpc_error should be returning errors for partial errors - i.e., for errors that are embedded in the regular response returned by VtGate RPC calls. I assume error.code only exists for when errors are returned without a response.

File: py/vtdb/grpc_vtgate_client.py [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+        rowset = _get_rowset_from_query_result(result)
+        rowsets.append(rowset)
+      return rowsets
+
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      self.logger_object.log_private_data(bind_variables_list)
+      raise _convert_exception(
+          e, sql_list, exec_method, keyspace='', tablet_type=tablet_type)
+
+  def _stream_execute(
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: I believe you want to put a retry decorator around the execute methods.

File: py/vtdb/grpc_vtgate_client.py [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+    for field in first_response.result.fields:
+      fields.append((field.name, field.type))
+      conversions.append(field_types_proto3.conversions.get(field.type))
+
+    def row_generator():
+      try:
+        for response in it:
+          for row in response.result.rows:
+            yield tuple(_make_row(row, conversions))
+      except Exception:
```
Comment (aaijazi) [inline py/vtdb/grpc_vtgate_client.py]: Any way we can pare this down to not catch all exceptions?

File: py/vtdb/vtgate_utils.py [hunk 9]
```diff
@@ -117,6 +117,9 @@ def convert_to_dbexception(self, args):
     Returns:
       An exception from dbexceptions.
     """
+    # FIXME(alainjobart): this is extremely confusing: self.message is only
```
Comment (aaijazi) [inline py/vtdb/vtgate_utils.py]: Yeah, to be honest, I'm not sure why we handle integrity error message pruning differently from all other errors.

**Comment by aaijazi** [comment]
Generally looks good, nice cleanup. gRPC certainly makes some stuff a lot simpler than GoRPC (for example: streaming).

I left a few comments, and should talk in person about errors this week.

File: py/vtdb/vtgate_utils.py (hunk 9 above)
Comment (dumbunny) [inline py/vtdb/vtgate_utils.py]: PII. Integrity errors print the values clause of insert statements.

**Pull request merged**